flink-connector-dynamodb/src/main/java/org/apache/flink/streaming/connectors/dynamodb/util/PrimaryKeyBuilder.java

```
@@ -0,0 +1,104 @@
+import org.apache.flink.util.CollectionUtil;
+
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+import software.amazon.awssdk.services.dynamodb.model.WriteRequest;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/** Helper to construct primary(composite) key for a DynamoDB request. */
+@Internal
+public class PrimaryKeyBuilder {
+
+    public static String build(List<String> pKeys, WriteRequest request) {
+
+        if (CollectionUtil.isNullOrEmpty(pKeys)) {
+            // fake key, because no dynamodb partition key or sort key provided. Using UUID should be safe
+            // here, as we have at most 25 elements per batch
+            return UUID.randomUUID().toString();
+        } else {
+            Map<String, AttributeValue> requestItems = getRequestItems(request);
+
+            StringBuilder builder = new StringBuilder();
+            for (String keyName : pKeys) {
+                AttributeValue value = requestItems.get(keyName);
+
+                if (value == null) {
+                    throw new InvalidRequestException(
+                            "Request " + request.toString() + " does not contain pKey " + keyName);
+                }
+
+                builder.append(getKeyValue(value));
+            }
+
+            return builder.toString();
+        }
+    }
+
+    /**
+     * Returns string value of a partition key attribute. Each primary key attribute must be defined
+     * as type String, Number, or binary as per DynamoDB specification.
+     */
+    private static String getKeyValue(AttributeValue value) {
+        StringBuilder builder = new StringBuilder();
+
+        if (value.n() != null) {
+            builder.append(value.n());
+        }
+
+        if (value.s() != null) {
+            builder.append(value.s());
+        }
+
+        if (value.b() != null) {
+            builder.append(value.b().asUtf8String());
+        }
+
+        return builder.toString();
```

Do we want to validate that no unsupported types were used, or is this done elsewhere?

---

DynamoDB only allows string, number or binary as key. We can do additional validation, but that would be applied on every item. Do you think it worth having this additional check given the performance hit?

---

We already validate that the `AttributeValue` is not null above. How about we validate the return of this method is not empty?

---

I added validation as per DynamoDB specification partition key, and sort key values can not be blank. However, we would only be able to validate that if the deduplication configuration is provided. When not provided, the request will be failed by DynamoDB.

DynamoDB would vail request if we do not validate for the blank values at all, but I agree it is good to fail fast.